features/createroom/impl/src/main/kotlin/io/element/android/features/createroom/impl/joinbyaddress/JoinRoomByAddressPresenter.kt

```diff
@@ -0,0 +1,138 @@
+                is JoinRoomByAddressEvents.UpdateAddress -> {
+                    validateAddress = false
+                    address = event.address.trim()
+                }
+            }
+        }
+
+        RoomAddressStateEffect(
+            fullAddress = address,
+            onRoomAddressStateChange = { addressState ->
+                internalAddressState = addressState
+                if (addressState is RoomAddressState.RoomFound && validateAddress) {
+                    onRoomFound(addressState)
+                }
+            }
+        )
+
+        val addressState by remember {
+            derivedStateOf {
+                // We only want to show the "RoomFound" state as long as the user didn't validate the address.
+                if (validateAddress || internalAddressState is RoomAddressState.RoomFound) {
+                    internalAddressState
+                } else {
+                    RoomAddressState.Unknown
+                }
+            }
+        }
+
+        return JoinRoomByAddressState(
+            address = address,
+            addressState = addressState,
+            eventSink = ::handleEvents
+        )
+    }
+
+    private fun onRoomFound(state: RoomAddressState.RoomFound) {
+        navigator.onDismissJoinRoomByAddress()
+        navigator.onOpenRoom(
+            roomIdOrAlias = state.resolved.roomId.toRoomIdOrAlias(),
+            serverNames = state.resolved.servers
+        )
+    }
+
+    @Composable
+    private fun RoomAddressStateEffect(
+        fullAddress: String,
+        onRoomAddressStateChange: (RoomAddressState) -> Unit,
+    ) {
+        val onChange by rememberUpdatedState(onRoomAddressStateChange)
+        LaunchedEffect(fullAddress) {
+            // Whenever the address changes, reset the state to unknown
+            onChange(RoomAddressState.Unknown)
+            // debounce the room address resolution
+            delay(300)
+            val roomAlias = tryOrNull { RoomAlias(fullAddress) }
+            if (roomAlias != null && roomAliasHelper.isRoomAliasValid(roomAlias)) {
+                onChange(RoomAddressState.Resolving)
+                onChange(client.resolveRoomAddress(roomAlias))
+            } else {
+                onChange(RoomAddressState.Invalid)
+            }
+        }
+    }
+
+    private suspend fun MatrixClient.resolveRoomAddress(roomAlias: RoomAlias): RoomAddressState {
+        return withTimeoutOrNull(ADDRESS_RESOLVE_TIMEOUT_IN_SECONDS.seconds) {
+            resolveRoomAlias(roomAlias)
+                .fold(
+                    onSuccess = { resolved ->
+                        if (resolved.isPresent) {
+                            RoomAddressState.RoomFound(resolved.get())
+                        } else {
+                            RoomAddressState.RoomNotFound
+                        }
+                    },
+                    onFailure = { _ -> RoomAddressState.RoomNotFound }
```

Comment:
We should maybe handle network error case separately?

Reply:
For now I'll leave it like this, will see with design/product, but I think it's ok